numpyro/infer/util.py

```diff
@@ -303,30 +333,43 @@ def initialize_model(rng_key, model,
         `potential_fn` and `constraints_fn` callables, respectively.
     :param tuple model_args: args provided to the model.
     :param dict model_kwargs: kwargs provided to the model.
-    :return: tuple of (`init_params`, `potential_fn`, `postprocess_fn`),
-        `init_params` are values from the prior used to initiate MCMC,
+    :return: tuple of (`init_params_info`, `potential_fn`, `postprocess_fn`, `model_trace`),
+        `init_params_info` is a tuple containing values from the prior used to initiate MCMC,
+        their corresponding potential energy, and their gradients,
         `postprocess_fn` is a callable that uses inverse transforms
         to convert unconstrained HMC samples to constrained values that
         lie within the site's support, in addition to returning values
         at `deterministic` sites in the model.
     """
-    if model_kwargs is None:
-        model_kwargs = {}
-    potential_fn, postprocess_fn = get_potential_fn(rng_key if rng_key.ndim == 1 else rng_key[0],
-                                                    model,
+    model_kwargs = {} if model_kwargs is None else model_kwargs
+    inv_transforms, replay_model, model_trace = get_model_transforms(
+        model, model_args, model_kwargs)
+    constrained_values = {k: v['value'] for k, v in model_trace.items()
+                          if v['type'] == 'sample' and not v['is_observed']}
+
+    potential_fn, postprocess_fn = get_potential_fn(model,
+                                                    inv_transforms,
+                                                    replay_model=replay_model,
                                                     dynamic_args=dynamic_args,
                                                     model_args=model_args,
                                                     model_kwargs=model_kwargs)
 
-    init_params, is_valid = find_valid_initial_params(rng_key, model,
-                                                      init_strategy=init_strategy,
-                                                      model_args=model_args,
-                                                      model_kwargs=model_kwargs)
+    init_strategy = init_strategy if isinstance(init_strategy, partial) else init_strategy()
+    if init_strategy.func is init_to_value:
+        init_values = init_strategy.keywords.get("values")
+        unconstrained_values = transform_fn(inv_transforms, init_values, invert=True)
+        init_strategy = _init_to_unconstrained_value(values=unconstrained_values)
+    prototype_params = transform_fn(inv_transforms, constrained_values, invert=True)
```

Comment:
If the init strategy is `init_to_value`, we transform it into the unconstrained domain.